pkg/apis/extensions/validation/validation.go

```diff
@@ -828,26 +828,47 @@ func ValidateNetworkPolicySpec(spec *extensions.NetworkPolicySpec, fldPath *fiel
 	allErrs = append(allErrs, unversionedvalidation.ValidateLabelSelector(&spec.PodSelector, fldPath.Child("podSelector"))...)
 
 	// Validate ingress rules.
-	for _, i := range spec.Ingress {
+	for i, ingress := range spec.Ingress {
+		ingressPath := fldPath.Child("ingress").Index(i)
+		for i, port := range ingress.Ports {
+			portPath := ingressPath.Child("ports").Index(i)
+			protocol := ""
+			if port.Protocol == nil {
+				protocol = string(api.ProtocolTCP)
+			} else if *port.Protocol == api.ProtocolTCP || *port.Protocol == api.ProtocolUDP {
+				protocol = string(*port.Protocol)
+			} else {
+				allErrs = append(allErrs, field.NotSupported(portPath.Child("protocol"), *port.Protocol, []string{string(api.ProtocolTCP), string(api.ProtocolUDP)}))
+			}
+			if port.Port != nil {
+				if port.Port.Type == intstr.Int {
+					for _, msg := range validation.IsValidPortNum(int(port.Port.IntVal)) {
+						allErrs = append(allErrs, field.Invalid(portPath.Child("port"), port.Port.IntVal, msg))
+					}
+				} else {
+					if _, err := net.LookupPort(strings.ToLower(protocol), port.Port.StrVal); err != nil {
```

Comment:
String portnames actually mean to look up that named port on Pod.  Yes, it means ports could vary across different pods.